CMakeLists.txt

```diff
@@ -0,0 +1,559 @@
+    endif()
+elseif(NOT MSVC)
+    if(EMSCRIPTEN)
+        set(CMAKE_C_FLAGS_RELEASE "${CMAKE_C_FLAGS_RELEASE} -Os -fdata-sections -ffunction-sections")
+        if("${CMAKE_C_COMPILER_ID}" STREQUAL "Clang")
+            set(LINK_FLAGS_RELEASE  "${LINK_FLAGS_RELEASE} -dead_strip")
+        else()
+            set(CMAKE_C_FLAGS_RELEASE "${CMAKE_C_FLAGS_RELEASE} -Wl,--gc-sections -Wl,-s")
+            set(LINK_FLAGS_RELEASE  "${LINK_FLAGS_RELEASE} -Wl,--gc-sections -Wl,-s")
+        endif()
+    else()
+        string(REGEX REPLACE "-O3" ""
+            CMAKE_C_FLAGS_RELEASE "${CMAKE_C_FLAGS_RELEASE}")
+        set(CMAKE_C_FLAGS_RELEASE "${CMAKE_C_FLAGS_RELEASE} -O2 -fdata-sections -ffunction-sections")
+        if(ANDROID)
+            set(CMAKE_C_FLAGS_DEBUG "${CMAKE_C_FLAGS_DEBUG} -funwind-tables")
+        elseif(NOT "${CMAKE_C_COMPILER_ID}" STREQUAL "Clang")
+            set(CMAKE_C_FLAGS_RELEASE "${CMAKE_C_FLAGS_RELEASE} -s -Wl,--gc-sections -Wl,-s")
+            set(LINK_FLAGS_RELEASE  "${LINK_FLAGS_RELEASE} -Wl,--gc-sections -Wl,-s")
+        else()
+            set(LINK_FLAGS_RELEASE  "${LINK_FLAGS_RELEASE} -dead_strip")
+        endif()
+    endif()
+endif()
+
+# Global optimization flags
+if(NOT MSVC)
+    set(CMAKE_C_FLAGS_RELEASE "${CMAKE_C_FLAGS_RELEASE} -fno-omit-frame-pointer")
+endif()
+
+if(NOT CMAKE_BUILD_TYPE)
+    set(CMAKE_BUILD_TYPE "Release")
+endif()
+
+string(TOLOWER ${CMAKE_BUILD_TYPE} CMAKE_BUILD_TYPE_LOWER)
+
+if(CMAKE_BUILD_TYPE_LOWER STREQUAL "release")
+    add_definitions(-DNDEBUG)
+endif()
+
+if(MSVC)
+    # Force to always compile with W4
+    if(CMAKE_C_FLAGS MATCHES "/W[0-4]")
+        string(REGEX REPLACE "/W[0-4]" "/W4" CMAKE_C_FLAGS "${CMAKE_C_FLAGS}")
+    else()
+        set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} /W4")
+    endif()
+elseif(CMAKE_COMPILER_IS_GNUCC OR CMAKE_COMPILER_IS_GNUCXX)
+    # Update if necessary
+    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -Wall -pedantic")
+endif()
+
+# Disable bogus MSVC warnings
+if(MSVC)
+    add_definitions(-D_CRT_SECURE_NO_WARNINGS)
+endif()
+
+# -fPIC thing
+if(NOT WIN32)
+    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -fPIC")
+endif()
+
+# Prevent shared libraries has "lib" prefix on Windows DLL files
+if(WIN32)
+    set(CMAKE_SHARED_LIBRARY_PREFIX "")
+endif()
+# -----------------------------------
+
+
+if(POLICY CMP0077) # Cache variables override since 3.12
+    cmake_policy(SET CMP0077 NEW)
+endif()
+
+if(EMSCRIPTEN)
+    set(BUILD_SHARED_DEFAULT OFF)
+else()
+    set(BUILD_SHARED_DEFAULT ON)
+endif()
+
+option(BUILD_STATIC   "Build static library of libXMP" ON)
+option(BUILD_SHARED   "Build shared library of libXMP" ${BUILD_SHARED_DEFAULT})
+if(NOT BUILD_STATIC AND NOT BUILD_SHARED)
+    message(FATAL_ERROR "Both static and shared builds got disabled. You must enable at least one of them.")
+endif()
+
+option(LIBXMP_BUILD_LITE   "Build lite version of libXMP" ON)
+option(LIBXMP_BUILD_FULL   "Build full version of libXMP" ON)
```

Comment:
I think this is the wrong thing to do:

Two release distributions are generated from the git tree: the 'normal' one
and the 'lite' one, and they do _not_ mix in terms of build system.  I.e. there
should be two `CMakeLists.txt` files, one at top-level for 'normal' version
and another under 'lite' directory. You can always keep the common stuff
in an include file or something. To see the needed differences, see configure.ac
and Makefile.in differences between the normal and the lite versions.

Reply:
I made the mixed build to simplify the whole building of both from one repository, the current build of lite looks too hacky and not so nice to build it, but yeah, I can split the thing into include files and two separated CMake files that do build full and lite versions but keep shared fragments being same.

Reply:
The cmake thing is supposed to work in the released versions:
with the current release method, your existing cmake'ry _shall_ fail.

Reply:
Right now my CMake build works in a case to build both full and lite versions and install them into the system. BUT, when you speak about the tarball packing, that right, as my CMake method wasn't expected to have separate builds for the full and lite versions, allowing me to install both (and then, choose which of them to link and use at the software).

I'll try to perform the CMake build split tomorrow, as right now I am tired and it's too late time (3:26 UTC+3).

Reply:
A little question for the future (after I fix the MSVC unit test freeze problem): how I could perform the split into two CMake builds for full and lite configs?
1) The Full config can build both full and lite libraries; Lite config will build the lite only library
2) The Full config will build the full only library; Lite config will build the lite only library
Also a third option for the full config:
3) Same as 1, however, Lite version is disabled by default at the Full config, but can be enabled by the user if needed

Reply:
It's 2, without a third option

Reply:
Ok, will do this once I'll get this step to do :+1: